src/node_report_module.cc

```diff
@@ -0,0 +1,327 @@
+#include "env.h"
+#include "node_errors.h"
+#include "node_internals.h"
+#include "node_options.h"
+#include "node_report.h"
+#include "util.h"
+
+#include "env-inl.h"
+#include "handle_wrap.h"
+#include "node_buffer.h"
+#include "stream_base-inl.h"
+#include "stream_wrap.h"
+#include "util-inl.h"
+
+#include <sstream>
+#include <atomic>
+#include <v8.h>
+
+namespace nodereport {
+using node::Environment;
+using node::FIXED_ONE_BYTE_STRING;
+using node::IsolateData;
+using node::PerIsolateOptions;
+using node::Utf8Value;
+using v8::Array;
+using v8::Boolean;
+using v8::Context;
+using v8::Function;
+using v8::FunctionCallbackInfo;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::MaybeLocal;
+using v8::Object;
+using v8::String;
+using v8::V8;
+using v8::Value;
+
+// Internal/static function declarations
+static void OnFatalError(const char* location, const char* message);
+void OnUncaughtException(const FunctionCallbackInfo<Value>& info);
+static void Initialize(Local<Object> exports,
+                       Local<Value> unused,
+                       Local<Context> context);
+
+/*******************************************************************************
+ * External JavaScript API for triggering a report
+ *
+ ******************************************************************************/
+void TriggerReport(const FunctionCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+  std::string filename;
+  MaybeLocal<Value> error;
+
+  if (info.Length() == 1) {
+    error = info[0];
+  } else {
+    filename = *String::Utf8Value(isolate, info[0]);
+    error = info[1];
+  }
+
+  TriggerNodeReport(isolate, env, "JavaScript API", __func__, filename, error);
+  // Return value is the report filename
+  info.GetReturnValue().Set(String::NewFromUtf8(isolate, filename.c_str()));
+}
+
+/*******************************************************************************
+ * External JavaScript API for returning a report
+ ******************************************************************************/
+void GetReport(const FunctionCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+  std::ostringstream out;
+
+  GetNodeReport(isolate, env, "JavaScript API", __func__, info[0], out);
+
+  // Return value is the contents of a report as a string.
+  info.GetReturnValue().Set(String::NewFromUtf8(isolate, out.str().c_str()));
```

Comment:
Is there ever a chance that a null byte could end up in `out`?

Reply:
I don't think so; the calee APIs have logic to write data into the stream unconditionally, so null bytes are impossible with the given control flow.